solidity/contracts/ValidatorRegistration.sol

```diff
@@ -1,23 +1,113 @@
+    )
+        public
+        payable
+    {
+        uint index = totalDepositCount + 2 ** MERKLE_TREE_DEPTH;
+        bytes memory msgGweiInBytes = toBytes(msg.value);
+        bytes memory timeStampInBytes = toBytes(block.timestamp);
+        bytes memory depositData = mergeBytes(mergeBytes(msgGweiInBytes, timeStampInBytes), depositParams);
 
-}
+        emit Eth1Deposit(receiptTree[1], depositParams, totalDepositCount);
+
+        receiptTree[index] = abi.encodePacked(keccak256(depositData));
+        for (uint i = 0; i < MERKLE_TREE_DEPTH; i++) {
+            index = index / 2;
+            receiptTree[index] = abi.encodePacked(keccak256(mergeBytes(receiptTree[index * 2], receiptTree[index * 2 + 1])));
+        }
+
+        require(
+            msg.value <= DEPOSIT_SIZE,
+            "Deposit can't be greater than DEPOSIT_SIZE."
+        );
+        require(
+            msg.value >= MIN_TOPUP_SIZE,
+            "Deposit can't be lesser than MIN_TOPUP_SIZE."
+        );
+        if (msg.value == DEPOSIT_SIZE) {
+            totalDepositCount++;
+        }
+
+        // When ChainStart log publishes, beacon chain node initializes the chain and use timestampDayBoundry
+        // as genesis time.
+        if (totalDepositCount == DEPOSITS_FOR_CHAIN_START) {
+            uint timestampDayBoundry = block.timestamp - block.timestamp % SECONDS_PER_DAY + SECONDS_PER_DAY;
+            bytes memory timestampDayBoundryBytes = toBytes(timestampDayBoundry);
+            emit ChainStart(receiptTree[1], timestampDayBoundryBytes);
+        }
+    }
+
+    function getReceiptRoot() public constant returns (bytes) {
+        return receiptTree[1];
+    }
+
+    function toBytes(uint x) private constant returns (bytes memory) {
+        bytes memory b = new bytes(32);
+        assembly {
+          mstore(add(b, 32), x)
+        }
+        return b;
+    }
+
+    function mergeBytes(bytes memory a, bytes memory b) private returns (bytes memory c) {
+        // Store the length of the first array
+        uint alen = a.length;
+        // Store the length of BOTH arrays
+        uint totallen = alen + b.length;
+        // Count the loops required for array a (sets of 32 bytes)
+        uint loopsa = (a.length + 31) / 32;
+        // Count the loops required for array b (sets of 32 bytes)
+        uint loopsb = (b.length + 31) / 32;
+        assembly {
+            let m := mload(0x40)
+            // Load the length of both arrays to the head of the new bytes array
+            mstore(m, totallen)
+            // Add the contents of a to the array
+            for {  let i := 0 } lt(i, loopsa) { i := add(1, i) } { mstore(add(m, mul(32, add(1, i))), mload(add(a, mul(32, add(1, i))))) }
+            // Add the contents of b to the array
+            for {  let i := 0 } lt(i, loopsb) { i := add(1, i) } { mstore(add(m, add(mul(32, add(1, i)), alen)), mload(add(b, mul(32, add(1, i))))) }
+            mstore(0x40, add(m, add(32, totallen)))
+            c := m
```

Comment:
the purpose of this function is to merge the two byte arrays into one byte array; it looks like here we are merging the two arrays one byte at a time. 

 would something like this work the same?
```
function merge(bytes memory a, bytes memory b) internal returns (bytes memory c) {
    assembly {
        let ptr := mload(0x40)
        /// might need to store a.length + b.length in first 32 bytes
        mstore(ptr, a)
        mstore(add(ptr, a.length), b)
        c := ptr
    }
}
```

Reply:
I believe what noot has is more efficient. Are there any solidity implementations we can reference? Perhaps from lighthouse, min, etc...

Reply:
I agree. Alot of the contract's logic came from [this file](https://github.com/prysmaticlabs/prysm/blob/master/contracts/validator-registration-contract/validator_registration.sol) in Prysm's codebase, so it may be better to confer with one of them regarding assembly. @noot @GregTheGreek 

I need to read more about Sol assembly before I comment on this.

Reply:
@GregTheGreek actually, I believe `mstore` only stores 32 bytes at a time, so if the byte array is longer than that, we will need a loop. going to think about this for a bit, it might be fine.

Reply:
wouldn't it still be more efficient following your method for larger than 32? We could toss an IF/ELSE to loop for > 32?

Reply:
@GregTheGreek yes, `abi.encodePacked` also appears to do what we want but it slices the lengths off. I'm wondering if using `abi.encodePacked(a.length + b.length, a, b)` would work

Reply:
@noot @GregTheGreek  I removed `mergeBytes` in favor of `abi.encodePacked`. After looking at noot's previous message, I think we should ask someone who knows Vyper whether the `concat` function from the spec removes the lengths of byte arrays just as `abi.encodePacked`. If it does **not** remove array lengths, we should pack arguments with their lengths like noot said above.